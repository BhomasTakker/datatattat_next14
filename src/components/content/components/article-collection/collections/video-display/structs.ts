export const enum PlayerCollectionVariant {
	HorizontalScroll = "HorizontalScroll",
	VerticalScroll = "VerticalScroll",
}

export const enum PlayerSourceTypes {
	Youtube = "video/youtube",
	Vimeo = "video/vimeo",
	Mp4 = "video/mp4",
}

export enum PlayerAutoplayOptions {
	Play = "play",
	Muted = "muted",
	Any = "any",
}
