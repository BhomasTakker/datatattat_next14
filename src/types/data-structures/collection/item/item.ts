import { BaseInfo, Details, Avatar } from "../base";

export type CollectionItem = BaseInfo & {
	details?: Details;
	avatar?: Avatar;
};
