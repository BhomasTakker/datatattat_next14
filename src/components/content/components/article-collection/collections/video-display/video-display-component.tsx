"use client";

import { CollectionItem } from "@/types/data-structures/collection/item/item";
import { VideoPlayerContainer } from "../../../video-player/video-player-container";
import { Interaction } from "../../article/interaction/interactions";
import { InteractionsOptions } from "../../article/interaction/interactions-map";

import styles from "./video-display.module.scss";
import { useRef } from "react";
import Player from "video.js/dist/types/player";
import {
	PlayerAutoplayOptions,
	PlayerCollectionVariant,
	PlayerSourceTypes,
} from "./structs";
import { InViewComponent } from "@/components/ui/in-view/in-view";
import { articleMetaLoader, articleRenderer, articleTemplate } from "../utils";
import { WithData } from "@/components/ui/with-data/with-data";
import { BASE_PLAYER_OPTIONS } from "../../../video-player/options";

type VideoDisplayComponentProps = {
	articles: CollectionItem[];

	variant: PlayerCollectionVariant;
	sourceType: PlayerSourceTypes;
	autoplay: boolean;
};

export const VideoDisplayComponent = ({
	articles = [],
	variant,
	sourceType, // use youtube as an id and then do whatever required for each option
	autoplay,
}: VideoDisplayComponentProps) => {
	const playerRef = useRef<Player>(null);
	const firstArticle = articles[0];
	const src = firstArticle?.src;
	const template = articleTemplate(styles);

	if (!firstArticle) {
		return null;
	}

	// These in a separate file
	// We will ultimately need to pass or reference sourcetype
	// based on item as we may want variant sources in the same collection
	const onClickHnd = (item: CollectionItem) => {
		const video = playerRef.current;
		if (!video) {
			return;
		}
		// This works to effectively reset the error - if there is an error
		// @ts-expect-error null works but flags error / undefine ddoes not work...
		video.error(null);
		video.autoplay(true);
		video.src({
			src: item.src,
			type: sourceType,
		});
		video.poster(item.avatar?.src);
	};

	const options = {
		...BASE_PLAYER_OPTIONS,
		autoplay: autoplay ? PlayerAutoplayOptions.Any : false,
		poster: firstArticle.avatar?.src || "",
		sources: [
			{
				src,
				type: sourceType,
			},
		],
	};

	const containerClassName = `container${variant}`;
	const containerClass = styles[containerClassName];

	// We want to create the video initialisation object here?
	// If we are passing in event handlers then we may well want to

	return (
		// class variant
		<div className={containerClass} data-testid={variant}>
			<div className={styles.videoPlayer} data-testid="video-player-container">
				<VideoPlayerContainer
					styles={styles}
					playerRef={playerRef}
					options={options}
				/>
			</div>
			<ul className={styles.articles} data-testid="articles-list">
				{articles.map((item) => (
					<li key={item.src}>
						<InViewComponent
							options={{
								threshold: 0,
								triggerOnce: true,
							}}
							template={<div className={styles.template} />}
						>
							{/* Should be part of a 'wider' Article Component  */}
							<Interaction
								key={item.title}
								type={InteractionsOptions.Click}
								onClick={() => onClickHnd(item)}
							>
								<WithData
									getter={articleMetaLoader(item)}
									callback={articleRenderer(styles)}
									template={template}
								/>
							</Interaction>
						</InViewComponent>
					</li>
				))}
			</ul>
		</div>
	);
};
