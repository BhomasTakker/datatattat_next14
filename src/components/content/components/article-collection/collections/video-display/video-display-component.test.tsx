import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { VideoDisplayComponent } from "./video-display-component";
import { CollectionItem } from "@/types/data-structures/collection/item/item";
import { PlayerCollectionVariant, PlayerSourceTypes } from "./structs";

// Mock child components
jest.mock("../../../video-player/video-player-container", () => ({
	VideoPlayerContainer: ({ playerRef, options }: any) => (
		<div
			data-testid="video-player-container"
			data-options={JSON.stringify(options)}
		/>
	),
}));
jest.mock("../../article/interaction/interactions", () => ({
	Interaction: ({ children, onClick }: any) => (
		<div data-testid="interaction" onClick={onClick}>
			{children}
		</div>
	),
}));
jest.mock("../../../../../../components/ui/in-view/in-view", () => ({
	InViewComponent: ({ children }: any) => (
		<div data-testid="in-view">{children}</div>
	),
}));
jest.mock("../../../../../../components/ui/with-data/with-data", () => ({
	WithData: ({ template }: any) => (
		<div data-testid="with-data">{template}</div>
	),
}));
jest.mock("../utils", () => ({
	articleMetaLoader: (item: any) => () => item,
	articleRenderer: () => () => <div data-testid="article-renderer" />,
	articleTemplate: () => <div data-testid="article-template" />,
}));

const mockArticles: CollectionItem[] = [
	{
		src: "video1.mp4",
		title: "Video 1",
		avatar: { src: "poster1.jpg" },
	} as CollectionItem,
	{
		src: "video2.mp4",
		title: "Video 2",
		avatar: { src: "poster2.jpg" },
	} as CollectionItem,
];

describe("VideoDisplayComponent", () => {
	it("renders the video player and articles", () => {
		const { container } = render(
			<VideoDisplayComponent
				articles={mockArticles}
				variant={PlayerCollectionVariant.HorizontalScroll}
				sourceType={PlayerSourceTypes.Youtube}
				autoplay={true}
			/>
		);
		expect(screen.getByTestId("video-player-container")).toBeInTheDocument();
		expect(screen.getAllByTestId("interaction")).toHaveLength(
			mockArticles.length
		);
		expect(screen.getAllByTestId("with-data")).toHaveLength(
			mockArticles.length
		);
		expect(container).toMatchSnapshot();
	});

	it("passes correct options to VideoPlayerContainer", () => {
		render(
			<VideoDisplayComponent
				articles={mockArticles}
				variant={PlayerCollectionVariant.HorizontalScroll}
				sourceType={PlayerSourceTypes.Youtube}
				autoplay={true}
			/>
		);
		const container = screen.getByTestId("video-player-container");
		const options = JSON.parse(container.getAttribute("data-options")!);
		expect(options.autoplay).toBe("any");
		expect(options.poster).toBe("poster1.jpg");
		expect(options.sources[0].src).toBe("video1.mp4");
		expect(options.sources[0].type).toBe(PlayerSourceTypes.Youtube);
	});

	it("renders nothing if articles is empty", () => {
		const { container } = render(
			<VideoDisplayComponent
				articles={[]}
				variant={PlayerCollectionVariant.HorizontalScroll}
				sourceType={PlayerSourceTypes.Youtube}
				autoplay={false}
			/>
		);
		// Should still render the container, but no articles
		expect(container.querySelectorAll("li")).toHaveLength(0);
	});

	it("calls onClick handler when an article is clicked", () => {
		render(
			<VideoDisplayComponent
				articles={mockArticles}
				variant={PlayerCollectionVariant.HorizontalScroll}
				sourceType={PlayerSourceTypes.Youtube}
				autoplay={false}
			/>
		);
		const interactions = screen.getAllByTestId("interaction");
		fireEvent.click(interactions[1]);
		// No error thrown, and handler is called (no-op in test)
		expect(interactions[1]).toBeInTheDocument();
	});

	it("applies the correct container class based on variant", () => {
		const { container } = render(
			<VideoDisplayComponent
				articles={mockArticles}
				variant={PlayerCollectionVariant.VerticalScroll}
				sourceType={PlayerSourceTypes.Youtube}
				autoplay={false}
			/>
		);

		const element = container.getElementsByClassName("containerVerticalScroll");
		expect(element.length).toBeGreaterThan(0);
		expect(container).toMatchSnapshot();
	});
});
