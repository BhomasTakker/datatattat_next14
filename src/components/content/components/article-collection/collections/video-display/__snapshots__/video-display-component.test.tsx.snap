// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDisplayComponent applies the correct container class based on variant 1`] = `
<div>
  <div
    class="containerVerticalScroll"
    data-testid="VerticalScroll"
  >
    <div
      class="videoPlayer"
      data-testid="video-player-container"
    >
      <div
        data-options="{"autoplay":false,"aspectRatio":"16:9","controls":true,"fill":true,"fluid":true,"responsive":true,"playbackRates":[0.25,0.5,0.75,1,1.25,1.5,1.75,2],"preload":"auto","techOrder":["youtube","html5"],"poster":"poster1.jpg","sources":[{"src":"video1.mp4","type":"video/youtube"}]}"
        data-testid="mock-video-player-container"
      />
    </div>
    <ul
      class="articles"
      data-testid="articles-list"
    >
      <li>
        <div
          data-testid="in-view"
        >
          <div
            data-testid="interaction"
          >
            <div
              data-testid="with-data"
            >
              <div
                data-testid="article-template"
              />
            </div>
          </div>
        </div>
      </li>
      <li>
        <div
          data-testid="in-view"
        >
          <div
            data-testid="interaction"
          >
            <div
              data-testid="with-data"
            >
              <div
                data-testid="article-template"
              />
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`VideoDisplayComponent renders the video player and articles 1`] = `
<div>
  <div
    class="containerHorizontalScroll"
    data-testid="HorizontalScroll"
  >
    <div
      class="videoPlayer"
      data-testid="video-player-container"
    >
      <div
        data-options="{"autoplay":"any","aspectRatio":"16:9","controls":true,"fill":true,"fluid":true,"responsive":true,"playbackRates":[0.25,0.5,0.75,1,1.25,1.5,1.75,2],"preload":"auto","techOrder":["youtube","html5"],"poster":"poster1.jpg","sources":[{"src":"video1.mp4","type":"video/youtube"}]}"
        data-testid="mock-video-player-container"
      />
    </div>
    <ul
      class="articles"
      data-testid="articles-list"
    >
      <li>
        <div
          data-testid="in-view"
        >
          <div
            data-testid="interaction"
          >
            <div
              data-testid="with-data"
            >
              <div
                data-testid="article-template"
              />
            </div>
          </div>
        </div>
      </li>
      <li>
        <div
          data-testid="in-view"
        >
          <div
            data-testid="interaction"
          >
            <div
              data-testid="with-data"
            >
              <div
                data-testid="article-template"
              />
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
